flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/EGBasedInputsLocationsRetrieverTest.java

```diff
@@ -0,0 +1,134 @@
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for {@link EGBasedInputsLocationsRetriever}.
+ */
+public class EGBasedInputsLocationsRetrieverTest extends TestLogger {
+
+	/**
+	 * Tests that can get the producers of consumed result partitions.
+	 */
+	@Test
+	public void testGetConsumedResultPartitionsProducers() throws Exception {
+		final JobVertex producer1 = ExecutionGraphTestUtils.createNoOpVertex(1);
+		final JobVertex producer2 = ExecutionGraphTestUtils.createNoOpVertex(1);
+		final JobVertex consumer = ExecutionGraphTestUtils.createNoOpVertex(1);
+		consumer.connectNewDataSetAsInput(producer1, DistributionPattern.ALL_TO_ALL, ResultPartitionType.PIPELINED);
+		consumer.connectNewDataSetAsInput(producer2, DistributionPattern.ALL_TO_ALL, ResultPartitionType.PIPELINED);
+
+		final ExecutionGraph eg = ExecutionGraphTestUtils.createSimpleTestGraph(new JobID(), producer1, producer2, consumer);
+		final EGBasedInputsLocationsRetriever inputsLocationsRetriever = new EGBasedInputsLocationsRetriever(eg);
+
+		ExecutionVertexID evIdOfProducer1 = new ExecutionVertexID(producer1.getID(), 0);
+		ExecutionVertexID evIdOfProducer2 = new ExecutionVertexID(producer2.getID(), 0);
+		ExecutionVertexID evIdOfConsumer = new ExecutionVertexID(consumer.getID(), 0);
+
+		Collection<Collection<ExecutionVertexID>> producersOfProducer1 =
+				inputsLocationsRetriever.getConsumedResultPartitionsProducers(evIdOfProducer1);
+		Collection<Collection<ExecutionVertexID>> producersOfProducer2 =
+				inputsLocationsRetriever.getConsumedResultPartitionsProducers(evIdOfProducer2);
+		Collection<Collection<ExecutionVertexID>> producersOfConsumer =
+				inputsLocationsRetriever.getConsumedResultPartitionsProducers(evIdOfConsumer);
+
+		assertThat(producersOfProducer1, is(empty()));
+		assertThat(producersOfProducer2, is(empty()));
+		assertThat(producersOfConsumer, contains(Arrays.asList(evIdOfProducer1), Arrays.asList(evIdOfProducer2)));
+	}
+
+	/**
+	 * Tests that when execution is not scheduled, getting task manager location will return null.
+	 */
+	@Test
+	public void testGetNullTaskManagerLocationIfNotScheduled() throws Exception {
+		final JobVertex jobVertex = ExecutionGraphTestUtils.createNoOpVertex(1);
+
+		final ExecutionGraph eg = ExecutionGraphTestUtils.createSimpleTestGraph(new JobID(), jobVertex);
+		final EGBasedInputsLocationsRetriever inputsLocationsRetriever = new EGBasedInputsLocationsRetriever(eg);
+
+		ExecutionVertexID executionVertexId = new ExecutionVertexID(jobVertex.getID(), 0);
+		Optional<CompletableFuture<TaskManagerLocation>> taskManagerLocation =
+				inputsLocationsRetriever.getTaskManagerLocation(executionVertexId);
+
+		assertFalse(taskManagerLocation.isPresent());
+	}
+
+	/**
+	 * Tests that it can get the task manager location in an Execution.
+	 */
+	@Test
+	public void testGetTaskManagerLocationWhenScheduled() throws Exception {
+		final JobVertex jobVertex = ExecutionGraphTestUtils.createNoOpVertex(1);
```

Comment:
The method `scheduleForExecution()` should be already considered legacy.

Imo, we should do something like:

```
		final JobVertex jobVertex = ExecutionGraphTestUtils.createNoOpVertex(1);

		final TestingLogicalSlot testingLogicalSlot = new TestingLogicalSlot();
		final ExecutionGraph eg = ExecutionGraphTestUtils.createSimpleTestGraph(new JobID(), jobVertex);
		final EGBasedInputsLocationsRetriever inputsLocationsRetriever = new EGBasedInputsLocationsRetriever(eg);

		final ExecutionVertex onlyExecutionVertex = eg.getAllExecutionVertices().iterator().next();
		onlyExecutionVertex.deployToSlot(testingLogicalSlot);

		ExecutionVertexID consumerVertexId = new ExecutionVertexID(jobVertex.getID(), 0);
		Optional<CompletableFuture<TaskManagerLocation>> taskManagerLocationOptional = inputsLocationsRetriever.getTaskManagerLocation(consumerVertexId);

		assertTrue(taskManagerLocationOptional.isPresent());

		final CompletableFuture<TaskManagerLocation> taskManagerLocationFuture = taskManagerLocationOptional.get();
		assertEquals(testingLogicalSlot.getTaskManagerLocation(), taskManagerLocationFuture.get());
```

What do you think?

Reply:
OK， reasonable